### DISCUSSION TITLE
Is this a good useSubscription hook?

### DISCUSSION THREAD

**Discussion by elliotwaite** [pmndrs/jotai#999] [Q&A] [answered]
I've been trying to come up with a `useSubscription` hook that could get called like this:
```
useSubscription(atom, (get, set, newVal) => {...})
```
Where the passed-in callback gets called whenever the atom updates, and the callback receives a get function, set function, and the new value of the atom. It's similar to a derived atom, but the callback also receives a set function, and any atoms that are read in the callback aren't tracked. — If something like this already exists, let me know. I searched, but couldn't find anything, so I tried to come up with my own.

This is what I've come up with so far (you can see a demo of it [here](https://codesandbox.io/s/jotai-usesubscription-demo-jp2p5?file=/src/App.jsx)):

```
function useSubscription(sourceAtom, callback) {
  const nonTrackingCallback = useAtomCallback(
    useCallback(
      (get, set, newVal) => {
        callback(get, set, newVal);
      },
      [callback]
    )
  );
  const trackingDerivedAtom = useMemo(
    () =>
      atom((get) => {
        // This get call is tracked, but any get calls in `callback` won't be.
        const newVal = get(sourceAtom);
        nonTrackingCallback(newVal);
      }),
    [sourceAtom, nonTrackingCallback]
  );
  useAtom(trackingDerivedAtom);
}
```
I'm new to Jotai, so I was hoping I could get some feedback on this technique (am I doing this correctly? could this be improved?). Also, let me know if a hook like this already exists.

**Comment by dai-shi** [comment, marked as answer]
Nice try, but this is violating a rule.
The `read` function must be pure. It's called in the React render phase. You can't call `set()` in render.
In concurrent rendering, it's really possible that `read` function is invoked more than once.

It might not be what you want, but if you run the callback in `useEffect`, your `useSubscription` api is possible. Let me know if you need help implementing it.

**Comment by elliotwaite** [comment]
[Edit: This should have been a reply to [Daishi's answer](https://github.com/pmndrs/jotai/discussions/999#discussioncomment-2118283) instead of a separate answer.]

Ah, I see. I'm a bit confused then. I thought I had to call `useAtom(trackingDerivedAtom)` outside of `useEffect` to register `trackingDerivedAtom`'s `read` function as a listener of `sourceAtom`, but if that's the case, I'm not sure how to also call the callback from within a `useEffect`. Can you elaborate on the `useEffect` suggestion?

I'm basically trying to implement a Jotai version of Valtio's `subscribe` function, such as how it's used in this example: [Subscribe and set local state conditionally](https://github.com/pmndrs/valtio/wiki/How-to-avoid-rerenders-manually#subscribe-and-set-local-state-conditionally)

Is there a simpler way to do this, or am I on the right track?

**Reply by dai-shi** [reply]
If you just need to conditionally set local state, a derived atom or `selectAtom` util can be used.

If you want to subscribe to an atom outside React, it's not currently supported. (but we will try an experimental feature in #861.)

useEffect-version of useSubscription in my mind is something like this:
```js
const useAtomSubscription = (anAtom, callback) => {
  const nonTrackingCallback = useAtomCallback(useCallback((get, set, newVal) => {
    callback(get, set, newVal);
  }, [callback]));
  const [value] = useAtom(anAtom)
  useEffect((() => {
    nonTrackingCallback(value);
  }, [nonTrackingCallback, value])
}
```

**Reply by elliotwaite** [reply]
Ah, okay. I updated my previous demo to use this implementation ([demo with useEffect](https://codesandbox.io/s/jotai-useatomsubscription-demo-with-useeffect-qhy85?file=/src/App.jsx:1272-1616)), but now I get extra re-renders. The other counts re-render whenever the main count changes.

I realized I can prevent those re-renders by putting the `useEffect` in a separate component. I created a separate `AtomSubscription` component that calls the `useEffect` and always returns `null`, and then I added that component as a child ([demo with AtomSubscription component](https://codesandbox.io/s/jotai-atomsubscription-component-demo-tux45?file=/src/App.jsx:1276-1632)):

```
function AtomSubscription({ anAtom, callback }) {
  const nonTrackingCallback = useAtomCallback(
    useCallback(
      (get, set, newVal) => {
        callback(get, set, newVal);
      },
      [callback]
    )
  );
  const [value] = useAtom(anAtom);
  useEffect(() => {
    nonTrackingCallback(value);
  }, [nonTrackingCallback, value]);
  return null;
}

function OtherCounter({ initialCount }) {
  const localCountAtom = useRef(atom(initialCount)).current;
  const [localCount, setLocalCount] = useAtom(localCountAtom);

  const callback = useCallback(
    (get, set, newVal) => {
      if (get(enabledAtom) && newVal >= get(localCountAtom)) {
        set(localCountAtom, newVal + 1);
      }
    },
    [localCountAtom]
  );

  ...

  return (
    <div>
      <AtomSubscription anAtom={mainCountAtom} callback={callback} />
      ...
    </div>
  );
}
```

I'm not sure if this is a good strategy or if I'm violating a different rule now.

That experimental `createStore` feature you mentioned (https://github.com/pmndrs/jotai/issues/861) looks like it would allow for a much simpler solution, by just using `store.sub(...)`.

I've tried to use derived atoms, but for the specific use case in the demo, I haven't been able to come up with a solution that avoids extra re-renders. But I'd be interested in seeing a solution if someone can suggest one.

I tried reimplementing the same demo using Valtio and it seemed more straightforward, so I might just switch to using Valtio for now, but I'd still be interested in any discussion about this topic related to Jotai. And thank you for your replies.

P.S. - I realized I made my last reply a separate answer. My mistake. I've added an edit to the top of it to try to help.

**Reply by dai-shi** [reply]
You are on the right track. It doesn't seem to violate any rules. It's just how React works. `AtomSubscription` would be a nice hack.

Looking at your example, here's what I would do without `useSubscription` hook: https://codesandbox.io/s/jotai-useatomsubscription-demo-with-useeffect-forked-kc5hu?file=/src/App.jsx

**Reply by elliotwaite** [reply]
Thanks, that's a helpful example. I was initially thinking that I didn't want to have to change `mainCountAtom` because the use case I had in mind was that `mainCountAtom` would be some central state of the app, and I wanted to be able to create any number of other components that could dynamically interact with that state without having to rewrite any logic inside `mainCountAtom`. But your example gave me an idea. I slightly tweaked your code and extracted it into a util, `atomWithSubscriptions` ([demo](https://codesandbox.io/s/jotai-atomwithsubscriptions-0jy97?file=/src/App.jsx)):
```
function atomWithSubscriptions(initialValue) {
  const baseAtom = atom(initialValue);
  const listenersAtom = atom([]);
  const anAtom = atom(
    (get) => get(baseAtom),
    (get, set, arg) => {
      set(baseAtom, arg);
      const newVal = get(baseAtom);
      for (const callback of get(listenersAtom)) {
        callback(get, set, newVal);
      }
    }
  );
  const useSubscription = (callback, deps) => {
    const memoizedCallback = useCallback(callback, deps);
    const setListeners = useUpdateAtom(listenersAtom);
    useEffect(() => {
      setListeners((prev) => [...prev, memoizedCallback]);
      return () =>
        setListeners((prev) => prev.filter((c) => c !== memoizedCallback));
    }, [setListeners, memoizedCallback]);
  };
  return [anAtom, useSubscription];
}
```

Which can be used like this:
```
const [countAtom, useCountSubscription] = atomWithSubscriptions(0);

function NumberOfTimesCountWasEvenComponent() {
  const evenCountAtom = useRef(atom(0)).current;
  const [evenCount, setEvenCount] = useAtom(evenCountAtom);

  useCountSubscription(
    (get, set, newVal) => {
      if (newVal % 2 === 0) {
        setEvenCount((c) => c + 1);
      }
    },
    [setEvenCount]
  );

  return <>{evenCount}</>;
}

```

I wasn't sure if calling `const memoizedCallback = useCallback(callback, deps)` inside the hook was okay, or if the callback needed to be memoized before being passed in. Also, I wasn't sure about the name of the util, I was also considering `atomWithListeners` and then something like `const [countAtom, useCountListenerCallback] = atomWithListeners(0)`.

I think ideally I would want to be able to subscribe to any atom without having to change that atom, but this is an interesting alternative solution. Thank you for your help.

**Reply by dai-shi** [reply]
Looks good!

> I wasn't sure if calling `const memoizedCallback = useCallback(callback, deps)` inside the hook was okay, or if the callback needed to be memoized before being passed in.

I wouldn't do this because the callback can be already stable, and explicit use of `useCallback` eases detecting mis-usage with eslint-plugin-react-hooks.
It's how `useAtomCallback` does. (I know it's a bit verbose, DX-wise.)

> Also, I wasn't sure about the name of the util, I was also considering `atomWithListeners` and then something like `const [countAtom, useCountListenerCallback] = atomWithListeners(0)`.

`atomWithListeners` sounds better to me. (we already have `atomWithSubscription`, and "subscription" somehow implies outside React for me.)

Would you like to add your idea in the advanced recipes?
https://github.com/pmndrs/jotai/blob/main/docs/advanced-recipes/atom-creators.mdx

**Reply by elliotwaite** [reply]
> I wouldn't do this because the callback can be already stable, and explicit use of useCallback eases detecting mis-usage with eslint-plugin-react-hooks.

Ah, good points.

> Would you like to add your idea in the advanced recipes?

Sure. I started a PR, but I think I need some help with the TypeScript types. I mentioned more in the PR: https://github.com/pmndrs/jotai/pull/1001

**Reply by elliotwaite** [reply]
I created an alternative version that I think simplifies usage. In this version, `atomWithListeners` only returns an atom, and there is a separate `useAtomListener` hook so that the user doesn't have to keep track of a different hook for each atom. The arguments to `useAtomListener` are the atom you want to listen to and the callback function. If the atom specified was not created with `atomWithListeners`, an error is raised.

Here's a [demo](https://codesandbox.io/s/jotai-atomwithlisteners-useatomlistener-goorp?file=/src/atomWithListeners.ts). Let me know if you think I should update the recipe in the docs to something like this.

**Reply by dai-shi** [reply]
This looks good. Only a caveat is it doesn't work with multiple Providers.
Can you try putting in an atom?
```js
const listenersMapAtom = atom(new Map());
```
The point is a) We need to immutably update Map to follow React convention, and b) We should clean up the map entry if it's empty to avoid memory leaks. (and we might want to register listeners before using `atomWithListeners`?)

**Reply by elliotwaite** [reply]
Ah, right. I forgot about immutable updates. Here's the [updated demo](https://codesandbox.io/s/jotai-atomwithlisteners-useatomlistener-v2-jgtzy?file=/src/atomWithListeners.ts).

What's new:
* The listeners map is now an atom.
* The updates don't mutate state.
* I switched the listeners type from a set to an array. Reasons:
  * I thought arrays might be more performant than sets for immutable style updates (but let me know if this is wrong).
  * An array allows the same callbacks to be registered multiple times, but this could also be implemented with a counting map, but I thought an array would be more performant than a counting map for immutable style updates (but again let me know if this is wrong).
* I changed the name of the atom's TypeScript type from `AtomType` to `AtomWithListeners`.
* The `useListener` hook no longer explicitly checks if the atom passed in was created using `atomWithListeners`, but TypeScript will now check for this ([here's the error in the demo](https://codesandbox.io/s/jotai-atomwithlisteners-useatomlistener-v2-jgtzy?file=/src/App.tsx:958-978)). I made `AtomWithListeners` an opaque type by intersecting it with `{ $$AtomWithListeners: true }` (I saw this style used in the Valtio code for `AsRef`).

Let me know if anything should be changed.

**Reply by dai-shi** [reply]
It looks great.
Although it would be useful, the implementation is a bit complicated.
For advanced-recipes, I think the current version is more readable. What do you think?
(It might be actually good for a 3rd-party library, if you are interested.)

**Reply by elliotwaite** [reply]
I agree. Also, with `unstable_createStore` now available in v1.6.0, it might be a good idea to just point users to use that instead and even remove my `atomWithListeners` recipe to avoid confusion of which method should be used. I just tried reimplementing my previous demo using `unstable_createStore` instead and it works great ([demo](https://codesandbox.io/s/jotai-createstore-r80is?file=/src/App.tsx)).

But I still enjoyed trying to come up with my own implementation, and it was a good learning experience, so thanks for the help with it.

**Reply by dai-shi** [reply]
Okay, let's keep what we have in advanced-recipes now. It's useful and good for learning too. No need to remove it.

`unstable_createStore` is unstable (api not fixed) and it's trickier to use in multi provider scenario.

**Reply by elliotwaite** [reply]
Okay, sounds good.
